orttraining/orttraining/training_api/module.cc

```diff
@@ -2,12 +2,137 @@
+  }
+  return Status::OK();
+}
+Module::Module(const std::string& train_model_path_or_bytes,
+               std::unordered_map<std::string, std::shared_ptr<Parameter>>& parameters,
+               const std::optional<std::string>& eval_model_path_or_bytes) {
+  // create value copy as same params are passed to Optimizer constructor
+  parameters_ = parameters;
+
+  auto so = onnxruntime::SessionOptions();
+  std::unique_ptr<Environment> env;
+  ORT_THROW_IF_ERROR(Environment::Create(nullptr, env));
+  train_sess_ = std::make_unique<onnxruntime::InferenceSession>(so, *env);
+  ORT_THROW_IF_ERROR(train_sess_->Load(train_model_path_or_bytes));
+  ORT_THROW_IF_ERROR(train_sess_->Initialize());
+  if (eval_model_path_or_bytes.has_value()) {
+    std::shared_ptr<onnxruntime::Model> eval_model;
+    ORT_THROW_IF_ERROR(onnxruntime::Model::Load(eval_model_path_or_bytes.value(), eval_model,
+                                                nullptr, env->GetLoggingManager()->DefaultLogger()));
+    GetGraphInputOutputNames(eval_model->MainGraph(), eval_input_names_, eval_output_names_);
+    // TODO:: do validation on eval inputs and outputs: eg order of user inputs, weights
+    eval_sess_ = std::make_unique<onnxruntime::InferenceSession>(so, *env);
+    ORT_THROW_IF_ERROR(eval_sess_->Load(eval_model_path_or_bytes.value()));
+    ORT_THROW_IF_ERROR(eval_sess_->Initialize());
+  }
+  auto& train_sess_state = train_sess_->GetSessionState();
+
+  std::shared_ptr<onnxruntime::Model> train_model;
+  ORT_THROW_IF_ERROR(onnxruntime::Model::Load(train_model_path_or_bytes, train_model,
+                                              nullptr, env->GetLoggingManager()->DefaultLogger()));
+  GetGraphInputOutputNames(train_model->MainGraph(), train_input_names_, train_output_names_);
+
+  std::vector<std::string> param_input_names, grad_input_names, user_input_names;
+  std::string param_name;
+  for (auto input_name : train_input_names_) {
+    auto it = parameters_.find(input_name);
+    if (it != parameters_.end()) {
+      param_input_names.emplace_back(input_name);
+      weights_.push_back(it->second->Data());
+    } else if (GetParamNameFromGradient(input_name, param_name)) {
+      grad_input_names.emplace_back(input_name);
+      // create gradient buffer
+      // assert param_name is valid.
+      auto it = parameters_.find(param_name);
+      if (it != parameters_.end()) {
+        ORT_THROW_IF_ERROR(it->second->AllocateGrad(input_name, train_sess_state));
+        gradients_.push_back(it->second->Gradient());
+      } else {
+        // raise error here.
+      }
+    } else {
+      user_input_names.emplace_back(input_name);
+    }
+  }
+  train_input_names_ = user_input_names;
+  train_input_names_.insert(train_input_names_.end(), param_input_names.begin(), param_input_names.end());
+  train_input_names_.insert(train_input_names_.end(), grad_input_names.begin(), grad_input_names.end());
```

Comment:
May we assume that the inputs are already ordered correctly? We don't need to perform this re-ordering since this is the contract between the offline tooling step and the backend.

Reply:
We can make some hard assumptions like:
1. User is using the correct checkpoint.
2. User input model is generated correctly.
and maybe some others. We can explicitly mention them here. But I don't think we should be re-ordering the inputs here.

Reply:
we are not exactly reordering inputs. The graph is not modified in any way. This code just gathers the corresponding ortvalues of things we need to pass internally. This was added since the initial test model had some out of order inputs. This is a one time cost, so I would prefer to keep it here for robustness